### PR TITLE
FEATURE: Add Multi Modal Capabilities to Flowise

### PR DESCRIPTION
**Pull request by vinodkiran**

### CODE REVIEW

File: packages/components/nodes/multimodal/OpenAI/OpenAIVisionChain.ts [hunk 1]
```diff
@@ -0,0 +1,288 @@
+import { ICommonObject, INode, INodeData, INodeOutputsValue, INodeParams } from '../../../src/Interface'
```
Comment (HenryHengZJ) [inline packages/components/nodes/multimodal/OpenAI/OpenAIVisionChain.ts:1]: rename the file name to `OpenAIMultiModalChain.ts`?

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+ */
+export class VLLMChain extends BaseChain implements OpenAIVisionChainInput {
+    static lc_name() {
+        return 'VLLMChain'
+    }
+    prompt: BasePromptTemplate | undefined
+
+    inputKey = 'input'
+    outputKey = 'text'
+    imageUrls?: []
```
Comment (HenryHengZJ) [inline packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:40]: we should change the name to something like `uploads` with interface.
This is being sent from the UI:
```
data: item.data,
type: item.type,
name: item.name,
mime: item.mime
```

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            organization: this.openAIOrganization
+        }
+
+        this.client = new OpenAIClient(this.clientConfig)
+    }
+
+    async _call(values: ChainValues): Promise<ChainValues> {
+        const userInput = values[this.inputKey]
+
+        const vRequest: any = {
```
Comment (HenryHengZJ) [inline packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:84]: can we implement the interface, would love avoiding using `any` if possible

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            model: this.modelName,
+            temperature: this.temperature,
+            top_p: this.topP,
+            messages: []
+        }
+        if (this.maxTokens) vRequest.max_tokens = this.maxTokens
+        else vRequest.max_tokens = 1024
+
+        const userRole: any = { role: 'user' }
```
Comment (HenryHengZJ) [inline packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:93]: same here, there should be some interface from `openai` js  library

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        userRole.content = []
+        userRole.content.push({
+            type: 'text',
+            text: userInput
```
Comment (HenryHengZJ) [inline packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:97]: what will be the `userInput` when there is only voice recording?

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                        detail: this.imageResolution
+                    }
+                })
+            }
+        }
+        vRequest.messages.push(userRole)
+        if (this.prompt && this.prompt instanceof ChatPromptTemplate) {
+            let chatPrompt = this.prompt as ChatPromptTemplate
+            chatPrompt.promptMessages.forEach((message: any) => {
+                if (message instanceof SystemMessagePromptTemplate) {
```
Comment (HenryHengZJ) [inline packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:151]: should we also check instanceof `HumanMessagePromptTemplate`?

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+                            }
+                        ]
+                    })
+                }
+            })
+        }
+
+        let response
+        try {
+            // @ts-ignore
```
Comment (HenryHengZJ) [inline packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:167]: why the ts-ignore here?

File: packages/server/src/index.ts [hunk 8]
```diff
@@ -1531,6 +1588,27 @@ export class App {
...
+                    if (upload.type === 'file') {
+                        const filename = upload.name
+                        const dir = path.join(getUserHome(), '.flowise', 'gptvision', chatId)
+                        if (!fs.existsSync(dir)) {
+                            fs.mkdirSync(dir, { recursive: true })
+                        }
+                        const filePath = path.join(dir, filename)
+                        const splitDataURI = upload.data.split(',')
+                        const bf = Buffer.from(splitDataURI.pop() || '', 'base64')
+                        //TODO: check if file exists, what should we do if it exists?
```
Comment (HenryHengZJ) [inline packages/server/src/index.ts:1603]: if file exists, we should replace it

File: packages/server/src/index.ts [hunk 9]
```diff
@@ -1531,6 +1588,27 @@ export class App {
...
+                        fs.writeFileSync(filePath, bf)
+                        // don't need to store the file contents in chatmessage, just the filename and chatId
+                        upload.data = chatId
+                        upload.type = 'stored-file'
```
Comment (HenryHengZJ) [inline packages/server/src/index.ts:1607]: will there be other `type`? would be great if we can use ENUM for different types, if there is only 1 type, probably we can remove it?

File: packages/ui/src/views/chatmessage/ChatMessage.js [hunk 10]
```diff
@@ -313,8 +595,62 @@ export const ChatMessage = ({ open, chatflowid, isDialog }) => {
...
+                                <button type='button' onClick={() => onRecordingCancelled()}>
+                                    Ok.
+                                </button>
+                            </div>
+                        </div>
+                    )}
+                </Box>
+            )}
+            <div
+                onDragOver={handleDragOver}
```
Comment (HenryHengZJ) [inline packages/ui/src/views/chatmessage/ChatMessage.js:647]: Moved `onDragOver={handleDragOver}` to this `div`. The initial implementation of this doesn't work when expand chatbox, it caused the messages to not render properly: 
image

However, there are also issues moving to here. Unless we can find a way to solve this without spending too much time, otherwise we can remove this drag drop functionality

**Comment by HenryHengZJ** [comment]
Update after touchup:
image

image

File: packages/server/src/index.ts (hunk 8 above)
Comment (vinodkiran) [reply packages/server/src/index.ts:1618]: Resolved! that is the default behavior. updated comment to reflect the same

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts (hunk 7 above)
Comment (vinodkiran) [reply packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:167]: Typo. Removed.

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts (hunk 6 above)
Comment (vinodkiran) [reply packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:151]: Yes, and pushed changes.

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts (hunk 5 above)
Comment (vinodkiran) [reply packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:97]: None! Some additional code is now checked in.

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts (hunk 4 above)
Comment (vinodkiran) [reply packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:93]: Updated.

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts (hunk 3 above)
Comment (vinodkiran) [reply packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:84]: Updated.

File: packages/components/nodes/multimodal/OpenAI/VLLMChain.ts (hunk 2 above)
Comment (vinodkiran) [reply packages/components/nodes/multimodal/OpenAI/VLLMChain.ts:40]: Updated.

File: packages/components/nodes/multimodal/OpenAI/OpenAIVisionChain.ts (hunk 1 above)
Comment (vinodkiran) [reply packages/components/nodes/multimodal/OpenAI/OpenAIVisionChain.ts:1]: Fixed in a new commit

File: packages/components/nodes/chains/ConversationChain/ConversationChain.ts [hunk 11]
```diff
@@ -128,16 +132,31 @@ const prepareChatPrompt = (nodeData: INodeData) => {
...
+    const imageContent = addImagesToMessages(nodeData, options)
+
+    //TODO, this should not be any[], what interface should it be?
+    let promptMessages: any[] = [
         SystemMessagePromptTemplate.fromTemplate(prompt ? `${prompt}\n${systemMessage}` : systemMessage),
         new MessagesPlaceholder(memory.memoryKey ?? 'chat_history'),
         HumanMessagePromptTemplate.fromTemplate(`{${inputKey}}`)
-    ])
+    ]
+    if (imageContent.length > 0) {
```
Comment (HenryHengZJ) [inline packages/components/nodes/chains/ConversationChain/ConversationChain.ts:150]: instead of doing on every chain/agent, can we do it on `ChatOpenAI`?

File: packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts [hunk 12]
```diff
@@ -19,7 +19,7 @@ class ChatOpenAI_ChatModels implements INode {
     constructor() {
```
Comment (HenryHengZJ) [inline packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts:19]: is it possible that we override the method `_generate` method and just append the images to the `messages` param in the func

File: packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts [hunk 13]
```diff
@@ -152,6 +152,73 @@ class ChatOpenAI_ChatModels implements INode {
                 type: 'json',
                 optional: true,
                 additionalParams: true
+            },
+            {
+                label: 'Allow Image Uploads',
```
Comment (HenryHengZJ) [inline packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts:157]: Maybe have a description saying it will automatically use gpt-vision when image is uploaded in the chat

File: packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts [hunk 14]
```diff
@@ -152,6 +152,73 @@ class ChatOpenAI_ChatModels implements INode {
...
+                name: 'allowImageUploads',
+                type: 'boolean',
+                default: false,
+                optional: true
+            },
+            {
+                label: 'Allow Audio Uploads',
+                name: 'allowAudioUploads',
```
Comment (HenryHengZJ) [inline packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts:165]: why the need for both `allowSpeechToText` and `allowAudioUploads` ?

Comment (vinodkiran) [reply packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts:165]: speechToText to allow Microphone recording as input, audioUploads to upload existing mp3/wav/... to be used as input (similar to images)

Comment (HenryHengZJ) [reply packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts:165]: although its good to have more options, but I think its kinda confusing to end users. 99% of users prob just want something similar to alexa, voice assistant, which is the speech to text toggle only.

**Comment by treesheeptw** [comment]
Hi, it seems like only GPT-4 Vision will be added in this pull. Will Ollama+LLaVA be added in the future? Ollama itself already supports LLaVA, but the chain does not. I wish this can be added along.

File: packages/components/nodes/chains/LLMChain/LLMChain.ts [hunk 15]
```diff
@@ -129,6 +130,7 @@ class LLMChain_Chains implements INode {
         if (!this.outputParser && outputParser) {
             this.outputParser = outputParser
         }
+        injectChainNodeData(nodeData, options)
```
Comment (vinodkiran) [inline packages/components/nodes/chains/LLMChain/LLMChain.ts:133]: @HenryHengZJ review critically

File: packages/components/nodes/chains/ConversationChain/ConversationChain.ts [hunk 16]
```diff
@@ -128,16 +131,19 @@ const prepareChatPrompt = (nodeData: INodeData) => {
 
     if (finalText) systemMessage = `${systemMessage}\nThe AI has the following context:\n${finalText}`
 
-    const chatPrompt = ChatPromptTemplate.fromMessages([
+    //TODO, this should not be any[], what interface should it be?
```
Comment (HenryHengZJ) [inline packages/components/nodes/chains/ConversationChain/ConversationChain.ts:134]: interface should be `BasesMessage[]`?

**Comment by HenryHengZJ** [comment]
> Hi, it seems like only GPT-4 Vision will be added in this pull. Will Ollama+LLaVA be added in the future? Ollama itself already supports LLaVA, but the chain does not. I wish this can be added along.

we'll first roll out chatopenai first, and move on to ollama

File: packages/components/nodes/chatmodels/ChatOpenAI/FlowiseChatOpenAI.ts [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+                    let inputVar = prompt.inputVariables[0]
+                    let formattedValues: any = {}
+                    formattedValues[inputVar] = audioTrans
+                    lastMessage.content = await prompt.format(formattedValues)
+                }
+                messages[0].push(lastMessage)
+            }
+        }
+        const messageContent = addImagesToMessages(nodeData, optionsData)
+        if (messageContent) {
```
Comment (HenryHengZJ) [inline packages/components/nodes/chatmodels/ChatOpenAI/FlowiseChatOpenAI.ts:60]: should be `(messageContent && messageContent .length)`

File: packages/components/nodes/chatmodels/ChatOpenAI/FlowiseChatOpenAI.ts [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+            if (messages[0].length > 0 && messages[0][messages[0].length - 1] instanceof HumanMessage) {
+                const lastMessage = messages[0].pop()
+                if (lastMessage instanceof HumanMessage) {
+                    lastMessage.content = messageContent
```
Comment (HenryHengZJ) [inline packages/components/nodes/chatmodels/ChatOpenAI/FlowiseChatOpenAI.ts:64]: `messageContent` only contains image:
```
[
    type: 'image_url',
    image_url: {
       url: bf,
       detail: 'low'
    }
]
```

Correct one should include user message as well:
```
[
    {
        type: "text",
        text: "What is in this image?",
     },
     {
        type: "image_url",
        image_url: bf,
      }
]
```

File: packages/components/nodes/chatmodels/ChatOpenAI/FlowiseChatOpenAI.ts (hunk 17 above)
Comment (mokeyish) [reply packages/components/nodes/chatmodels/ChatOpenAI/FlowiseChatOpenAI.ts:60]: should be `(messageContent?.length)` ?

Comment (HenryHengZJ) [reply packages/components/nodes/chatmodels/ChatOpenAI/FlowiseChatOpenAI.ts:60]: @mokeyish yep even better

**Comment by 0xi4o** [comment]
@vinodkiran pushed a couple of UI fixes:
- error message when audio recording is not supported
- messages not autoscrolling to the bottom

File: packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts [hunk 19]
```diff
@@ -243,16 +243,19 @@ class ChatOpenAI_ChatModels implements INode {
                 throw new Error("Invalid JSON in the ChatOpenAI's BaseOptions: " + exception)
             }
         }
-        const model = new FlowiseChatOpenAI(obj, {
+
+        const model = new ChatOpenAI(obj, {
             baseURL: basePath,
```
Comment (HenryHengZJ) [inline packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts:248]: can we make sure `baseURL` and `baseOptions` still works

File: packages/components/src/utils.ts [hunk 20]
```diff
@@ -758,3 +758,10 @@ export const prepareSandboxVars = (variables: IVariable[]) => {
     }
     return vars
 }
+
+/**
+ * Prepare storage path
+ */
+export const getStoragePath = (): string => {
+    return process.env.BLOB_STORAGE_PATH ? path.join(process.env.BLOB_STORAGE_PATH) : path.join(getUserHome(), '.flowise', 'storage')
```
Comment (HenryHengZJ) [inline packages/components/src/utils.ts:766]: todo: add `BLOB_STORAGE_PATH` to `start.ts` and other places

Comment (HenryHengZJ) [reply packages/components/src/utils.ts:766]: addressed

File: packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts (hunk 19 above)
Comment (HenryHengZJ) [reply packages/components/nodes/chatmodels/ChatOpenAI/ChatOpenAI.ts:248]: addressed

**Comment by 0xi4o** [comment]
@vinodkiran @HenryHengZJ Made a couple of changes:
- Change the UI for Speech to text configuration
- Made messages in view message dialog consistent with internal chat

**Comment by HenryHengZJ** [comment]
related chat embedded [PR](https://github.com/FlowiseAI/FlowiseChatEmbed/pull/94)

**Comment by 0xi4o** [comment]
@HenryHengZJ @chungyau97 Issues have all been fixed. We can review and merge this. I think everything's good to go on @vinodkiran's end too.

File: packages/ui/src/ui-component/dropdown/AsyncDropdown.js [hunk 21]
```diff
@@ -105,7 +105,11 @@ export const AsyncDropdown = ({
         })()
 
         // eslint-disable-next-line react-hooks/exhaustive-deps
-    }, [])
+    }, [credentialNames])
+
+    useEffect(() => {
```
Comment (HenryHengZJ) [inline packages/ui/src/ui-component/dropdown/AsyncDropdown.js:110]: this triggers inifite loop

File: packages/ui/src/ui-component/input/Input.js [hunk 22]
```diff
@@ -48,6 +48,12 @@ export const Input = ({ inputParam, value, nodes, edges, nodeId, onChange, disab
         }
     }, [myValue])
 
+    useEffect(() => {
+        if (value) {
```
Comment (HenryHengZJ) [inline packages/ui/src/ui-component/input/Input.js:52]: why this changes here @0xi4o

File: packages/ui/src/ui-component/dropdown/AsyncDropdown.js (hunk 21 above)
Comment (0xi4o) [reply packages/ui/src/ui-component/dropdown/AsyncDropdown.js]: I've reverted this change. I added this because the component wasn't rerendering on prop change. But it turns out the `useEffect` is not needed. It works without it.

File: packages/ui/src/ui-component/input/Input.js (hunk 22 above)
Comment (0xi4o) [reply packages/ui/src/ui-component/input/Input.js]: This is same as the `AsyncDropdown`. It works without the `useEffect`.

**Comment by chungyau97** [comment]
@0xi4o, thanks for the fix for invalid characters in file `2024-02-21 20_43_24-Elon Musk - Elon Musk.pdf — Mozilla Firefox.png`
image

File: packages/ui/src/ui-component/dropdown/AsyncDropdown.js [hunk 23]
```diff
@@ -105,7 +105,7 @@ export const AsyncDropdown = ({
         })()
 
         // eslint-disable-next-line react-hooks/exhaustive-deps
-    }, [])
+    }, [credentialNames])
```
Comment (HenryHengZJ) [inline packages/ui/src/ui-component/dropdown/AsyncDropdown.js:108]: this causes problem as well.

Just put in a custom tool or assistant:

1.) 
image

You will see it constantly loading

2.) From the terminal, you can see the method gets called infinitely:
image

**Comment by chungyau97** [comment]
Another error where importing flows then turn on `Speech To Text` will remove file and mic logo.
image

Steps to reproduce error: 
1. Click `Add New`
2. Import `MultiModal chatflow`
3. Save `chatflow`
4. Turn on `Speech To Text`

**Comment by HenryHengZJ** [comment]
Another issue:

1.) Open OpenAI Whisper, put in credential:
image

2.) switch to assembly ai, you can see the openai credential there:
image

File: packages/ui/src/ui-component/dropdown/AsyncDropdown.js (hunk 23 above)
Comment (HenryHengZJ) [reply packages/ui/src/ui-component/dropdown/AsyncDropdown.js]: resolved

**Comment by HenryHengZJ** [comment]
> Another error where importing flows then turn on `Speech To Text` will remove file and mic logo. image
> 
> Steps to reproduce error:
> 
> 1. Click `Add New`
> 2. Import `MultiModal chatflow`
> 3. Save `chatflow`
> 4. Turn on `Speech To Text`

solved

**Comment by HenryHengZJ** [comment]
> Another issue:
> 
> 1.) Open OpenAI Whisper, put in credential: image
> 
> 2.) switch to assembly ai, you can see the openai credential there: image
> 
> I think that's because of the `credentialNames` in the `useEffect`, I've removed that as it caused infinite loop, but you were trying to put it there to prevent this scenario right

solved

File: packages/ui/src/ui-component/dialog/SpeechToTextDialog.js [hunk 24]
```diff
@@ -0,0 +1,345 @@
...
+                                        {inputParam.label}
+                                        {!inputParam.optional && <span style={{ color: 'red' }}>&nbsp;*</span>}
+                                        {inputParam.description && (
+                                            <TooltipWithParser style={{ marginLeft: 10 }} title={inputParam.description} />
+                                        )}
+                                    </Typography>
+                                </div>
+                                {inputParam.type === 'credential' && (
+                                    <CredentialInputHandler
+                                        key={speechToText[selectedProvider]?.credentialId}
```
Comment (HenryHengZJ) [inline packages/ui/src/ui-component/dialog/SpeechToTextDialog.js:275]: @0xi4o this is where you can specify a `key` so that a new element is rendered, hence you dont need to put `credentialNames` in the `useEffect`

File: packages/server/src/index.ts [hunk 25]
```diff
@@ -1090,9 +1143,48 @@ export class App {
...
+
+            if (fs.existsSync(filePath)) {
+                res.setHeader('Content-Disposition', contentDisposition(path.basename(filePath)))
+                streamFileToUser(res, filePath)
+            } else {
+                return res.status(404).send(`File ${req.body.fileName} not found`)
+            }
+        })
+
+        this.app.get('/api/v1/get-upload-path', async (req: Request, res: Response) => {
```
Comment (ocflowiseai) [inline packages/server/src/index.ts:1155]: Reminder to backport this route to the new way modularity express router PR - https://github.com/FlowiseAI/Flowise/pull/1806

**Comment by HermesMacedo** [comment]
Could there be an option to configure another audio or image model (selfhosted)?

**Comment by nitromir** [comment]
Awesome feature. It would be great to add it to the chat embed as well.
